pkg/gadget-service/instance-manager/manager.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2023-2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package instancemanager
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+)
+
+type Service interface {
+	GetOperatorMap() map[operators.DataOperator]*params.Params
+}
+
+// Manager manages running gadgets without connection based context. It can run gadgets in the background as
+// well as buffer and multiplex their output.
+type Manager struct {
+	// mu is to be used whenever a gadget is installed or a new client wants to attach to a gadget
+	mu              sync.Mutex
+	gadgetInstances map[string]*GadgetInstance
+	waitingRoom     sync.Map
+
+	gadgetDone chan bool
+
+	// asyncGadgetRunCreation tells the Manager whether it is completely in control of creating gadget
+	// runs, or if those are (also) externally managed, like through custom resources in a kubernetes environment
+	asyncGadgetRunCreation bool
+
+	runtime runtime.Runtime
+	store   Store
+
+	Service
+}
+
+func New(runtime runtime.Runtime, options ...Option) (*Manager, error) {
+	mgr := &Manager{
+		gadgetInstances: make(map[string]*GadgetInstance),
+		gadgetDone:      make(chan bool, 1),
+		runtime:         runtime,
+	}
+	for _, opt := range options {
+		err := opt(mgr)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if mgr.store == nil {
+		return nil, errors.New("no store provided")
+	}
+	return mgr, nil
+}
+
+func (p *Manager) Store() Store {
+	return p.store
+}
+
+// StopGadget cancels a running gadget, but leaves the results accessible
+func (p *Manager) StopGadget(id string) error {
+	log.Printf("stopping gadget %q", id)
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	gadgetInstance, ok := p.gadgetInstances[id]
+	if !ok {
+		return fmt.Errorf("gadget not found")
+	}
+	gadgetInstance.cancel()
+
+	// ensure that the gadget is stopped before returning
+	if gadgetInstance.state == stateRunning {
+		<-p.gadgetDone
```

Comment:
Cancelling the context _should_ make it return sooner or later - this could only be prevented from loops coming for example from WASM afaik. We don't have a solution for that, but IMHO that's not a problem of the instance manager, but a problem that we would have to solve in the runtime.

But yeah, you're right, the critical section should be minimized.